Adapt testcontainers to be testing framework agnostic

pilosus opened the pull request:
## About

Testing: Adapt "Testcontainers" implementation to `unittest`

## References

- GH-58
- https://github.com/crate/crash/pull/408

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

amotl reviewed (commented):
Dear Vitaly,

thank you very much for your contribution, that looks excellent. Please give us some time to thoroughly review all the details.

On CI, there may be hiccups regarding to container boot timing / availability. We know it is a hassle to get hold of those issues.

```
E       docker.errors.APIError: 500 Server Error for http+docker://localhost/v1.43/containers/a519443561406c3f60e6d0b429cfd071736191f0149c9f32cc9deb9262818adc/start: Internal Server Error ("driver failed programming external connectivity on endpoint testcontainers-cratedb (68f8422f339f8d6b40efd574ecb54861825d9ccdcbba069cfc96e4a48032833a): Bind for 0.0.0.0:4200 failed: port is already allocated")
```

-- https://github.com/crate-workbench/cratedb-toolkit/actions/runs/6950127360/job/18922807956#step:6:4850

With kind regards,
Andreas.

amotl commented:
We know it is a hassle to get hold of those issues. Let us know if you still have the capacity to look at those details. On this matter, I think we will have to properly configure the Docker healthcheck subsystem [^1]. I think it can use something along the lines of `curl localhost:4200/_sql` or similar -- I will have to look this up in other recipes.

https://github.com/crate-workbench/cratedb-toolkit/blob/037b8bb0d31f8c3f144d1097b619857256c48b3a/cratedb_toolkit/testing/testcontainers/cratedb.py#L98-L103

[^1]: Given that it will be available. But I think I recently stumbled upon corresponding integration with Testcontainers.

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,8 +56,13 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
     KEEPALIVE = asbool(os.environ.get("CRATEDB_KEEPALIVE", os.environ.get("TC_KEEPALIVE", False)))
+    CMD_OPTS = {
+        "discovery.type": "single-node",
+        "cluster.routing.allocation.disk.threshold_enabled": False,
```
This line accidentally slipped in from my personal workstation configuration, because, well, disk usage is always >80% here ;]. May I humbly ask you to remove it from this patch?

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -101,3 +135,65 @@ def _connect(self):
+        self.cratedb.start()
+        self.database = DatabaseAdapter(dburi=self.get_connection_url())
+
+    def finalize(self):
+        if self.cratedb:
+            self.cratedb.stop()
```
I called this `finalize` here, but I don't know if it was a good idea. I don't think generic pytest fixtures have a convention for "naming things" on this matter, but I will be happy to lean towards one.

What do you think about `setup`/`teardown`, or `start`/`stop`?

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -101,3 +135,65 @@ def _connect(self):
+    ]
+
+
+class CrateDBFixture:
+    """
+    A little helper wrapping Testcontainer's `CrateDBContainer` and
+    CrateDB Toolkit's `DatabaseAdapter`, agnostic of the test framework.
+    """
```
I appreciate very much bringing this into the generic `cratedb_toolkit.testing` module namespace, to make it a re-usable component for other packages.

amotl replied:
Talking about "naming things", if you have a better idea for the name, in order to better disambiguate from pytest's notion of "fixtures", because it is actually a testframework-agnostic adapter/wrapper around, well, DatabaseAdapter [^1] and CrateDBContainer.

[^1]: Also eventually to be renamed to `CrateDBClientAdapter` or something different.

amotl replied:
Would `CrateDBTestAdapter` be a more sensible name?

pilosus replied:
Given that we have `DatabaseAdapter`, I like the `CrateDBTestAdapter` name!

amotl replied:
Wonderful. Let us rename it on behalf of a subsequent patch.

amotl replied:
I see you already renamed it properly, thanks.

I proposed to change it later, because I feared this would add too much noise to this patch because the symbol would need to be changed at too many places. But, of course, it turns out that it already has been nicely decoupled from the test cases on behalf of the `cratedb` and `cratedb_service` pytest fixtures, so my fears were unfounded.

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -101,3 +135,65 @@ def _connect(self):
         wait_for_logs(self, predicate="o.e.n.Node.*started", timeout=MAX_TRIES)
+
+
+class TestDrive:
+    """
+    Use different schemas for storing the subsystem database tables, and the
+    test/example data, so that they do not accidentally touch the default `doc`
+    schema.
+    """
+
+    EXT_SCHEMA = "testdrive-ext"
+    DATA_SCHEMA = "testdrive-data"
+
+    RESET_TABLES = [
+        f'"{EXT_SCHEMA}"."retention_policy"',
+        f'"{DATA_SCHEMA}"."raw_metrics"',
+        f'"{DATA_SCHEMA}"."sensor_readings"',
+        f'"{DATA_SCHEMA}"."testdrive"',
+        f'"{DATA_SCHEMA}"."foobar"',
+        f'"{DATA_SCHEMA}"."foobar_unique_single"',
+        f'"{DATA_SCHEMA}"."foobar_unique_composite"',
+        # cratedb_toolkit.io.{influxdb,mongodb}
+        '"testdrive"."demo"',
+    ]
```
On the other hand, contrary to `CrateDBFixture`, this section is specific to the test suite for `cratedb_toolkit`, and is not meant to be shipped with `cratedb_toolkit.testing`.

Do you see a chance to decouple this and let it be configured in `tests/conftest.py`, maybe on behalf of just a bit more pytest .request / _conf / -fixture magic, but loosely coupled, so that there is no dependency path going from `cratedb_testing` to `tests/conftest.py`, and the configuration could be somehow elegantly inverted instead?

I am not sure if I am asking for too much here, or if you can follow my thoughts easily, or if the implementation would be too complicated. Please let me know.

pilosus replied:
If reset tables aren’t need in testing module and only needed in conftest, I can simply move them there. But in this case shall I still try to remove explicit import of testing/testcontainers? It’s needed for CrateDBFixture. I probably don’t see the whole picture, but utils module is still explicitly imported there, so I don’t quite get why it’s different for testing module

amotl replied:
> If reset tables aren’t need in testing module and only needed in conftest, I can simply move them there.

This specific set of tables are only meant to be reset on behalf of the cratedb-toolkit test suite, which is not being shipped as part of the package. On the other hand, `cratedb_toolkit.testing` intends to bundle generic testing helpers/utilities/fixtures.

In this spirit, you made the right choice to put `CrateDBFixture` there, but I think the table definition / test suite configuration itself, now excellently bundled into the `TestDrive` container class, should stay in `tests/conftest.py`. However, it can't be there in isolation, because it will need to be picked up by the generic `CrateDBFixture` in some way, because this one actually orchestrates the container lifecycle. Can you figure out a way to make that happen elegantly?

> But in this case shall I still try to remove explicit import of testing/testcontainers? It’s needed for CrateDBFixture.

```
# File: cratedb_toolkit/testing/testcontainers/cratedb.py
from cratedb_toolkit.testing.testcontainers.util import KeepaliveContainer, asbool
```

```
# File: tests/conftest.py
from cratedb_toolkit.testing.testcontainers.cratedb import CrateDBFixture
```

Are you referring to one of those? I think both are fine in general. It should be free for every module to use generic utilities from `cratedb_toolkit.testing`, but not the other way round, at least import-wise.

> I probably don’t see the whole picture, but utils module is still explicitly imported there, so I don’t quite get why it’s different for testing module

Are you referring to this import?

```
# File: cratedb_toolkit/testing/testcontainers/cratedb.py
from testcontainers.core.waiting_utils import wait_container_is_ready, wait_for_logs
```

I think it is also perfectly fine. You can pull in all desired utitilities _into_ `conftest.py`, but the tricky part will be to define the test suite database connectivity configuration (`TestDrive`) _there_, and let it be picked up / consumed by the generic `CrateDBFixture` to be used properly at runtime, because it can't "import" something from `tests/conftest.py`.

pilosus replied:
So, basically, I've made `CrateDBFixture.reset` method to take reset tables in as a parameter without hardcoding them anyhow. That means, `CrateDBFixture` can still live in the `cratedb_toolkit/testing/testcontainers/cratedb.py`, but the reset tables are defined in the `tests/contest.py` along with the fixtures that use `CrateDBFixture`. I hope this solves your concern with the loose coupling.

amotl replied:
Yeah this was exactly what I was dreaming up regarding separation of concerns. Thank you very much.

pilosus commented:
I'll address other points too. Need more time to finalize.

amotl commented:
>> Bind for 0.0.0.0:4200 failed: port is already allocated
>
> is probably because of the port 4200 being used in the CI this way:

Ah I see. That makes sense.

> I think I can either revert back to the random port binding, or simply use another port in the fixture (which is probably better that the former option)

I think I'd prefer the random port assignment again for now, because it seems to be the standard way Testcontainers is doing it. By keeping up that paradigm, one has to be rigid to exclusively pick up the DBURI provided by the test machinery, and "not just sloppily assume there will be something on localhost:4200 or another fixed port anyway" ;].

What I don't like about the random port assignment paradigm is that is usually a bit more complicated to re-use test containers across subsequent test runs, and so it is mostly forgotten to be implemented on behalf of corresponding test container managers.

However, in our little corner, there is the generic `TC_KEEPALIVE=true` mechanism, which will keep the container running, so there is actually no penalty for subsequent test cycles. Personally, I am keeping this container running, dedicatedly for test harness use, and I still have the freedom to whip up ad hoc instances of CrateDB on the vanilla ports 4200/5432 for other purposes. Restarting the CrateDB container on each test run is unbearable.

> I'll address other points too. Need more time to finalize.

Thank you very much. Please take your time.

pilosus commented:
@amotl I've come up with the compromise solution to the hardcoded/random port binding for the `testcontainers`: the ports are to be taken in a a dict mapping inside container port to the port on the host machine. If the host port is `None`, a random host port will be generated. This follows Docker SDK for Python [conventions](https://docker-py.readthedocs.io/en/stable/containers.html) carefully. Some reasonable default when no ports dict supplied are also given (4200 to a random port). See how it works in the [crash PR](https://github.com/crate/crash/pull/408).

Also, I've fixed other points you've given above. Added a bit more docs, prettified the code.
Please, let me know if I missed something.

amotl commented on tests/conftest.py:
```diff
@@ -72,17 +44,18 @@ def cratedb_service():
+    db.reset(tables=RESET_TABLES)
     yield db
-    db.finalize()
+    db.stop()
```
Excellent. This reads much better now. Thanks.

amotl commented on tests/conftest.py:
```diff
@@ -72,17 +44,18 @@ def cratedb_service():
     Provide a fresh canvas to each test case invocation, by resetting database content.
     """
-    cratedb_service.reset()
+    cratedb_service.reset(tables=RESET_TABLES)
```
Ahja, excellent!

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -101,3 +150,53 @@ def _connect(self):
+            for reset_table in tables:
+                self.database.connection.exec_driver_sql(f"DROP TABLE IF EXISTS {reset_table};")
+
+    def get_connection_url(self, *args, **kwargs):
+        """
+        Return a URL for SQLAlchemy DB engine
+        """
+        if self.cratedb:
+            return self.cratedb.get_connection_url(*args, **kwargs)
+        return None
+
+    @property
+    def http_url(self):
+        """
+        Return a URL for HTTP interface
+        """
+        return self.get_connection_url(dialect="http")
```
#### Re.: Naming things

`get_connection_url` and `http_url` may also be renamed. However, it is also not relevant for this patch, but merely meant to encourage you that you can easily come up with ideas for renaming, because `cratedb-toolkit` is very young, and most of its code didn't see more than two or three iterations yet, so it offers many opportunities in this regard.

pilosus replied:
When naming these, we can go in two directions:
1. rely on the `testcontainers` naming conventions, e.g. if it's a connection URL that we get using `testcontainers.core.generic.DbContainer._create_connection_url` and the interface sugests to implement `get_connection_url`, then `get_connection_url` seems very natural to me.
2. rely on `CrateDB` naming. In this case we can say it's getting the [host](https://cratedb.com/docs/crate/crash/en/latest/run.html) url or the [interface](https://cratedb.com/docs/crate/reference/en/5.5/interfaces/index.html#) url. So that we can rename `get_connection_url` to `get_interface_url` or `get_host_url`

Tbh, I think the current naming is ok. If someone needs to go deeper into `testcontainers` base classes, `get_connection_url` looks better to me.

pilosus replied:
I'd rename `http_url` though to `get_http_url` to make it consistent with the `get_connection_url`.

amotl replied:
I agree with your assessment to follow the naming conventions of Testcontainers. Maybe we can have an inline docstring comment on the `get_connection_url` method to tell the readers that it is actually an SQLAlchemy database connection uri that will be returned here?

pilosus replied:
`get_connection_url` docstring says: `Return a URL for SQLAlchemy DB engine`. Do you want it to put it differently anyhow?

amotl replied:
Ah. No it is perfect. Apologies if I missed it.

amotl reviewed (approved):
Thank you again. Approved 💯. I will also add @seut and @matriv as secondary reviewers this time.

matriv commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,87 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
+        :param ports: optional dict that maps a port inside the container to a port on the host machine;
+                      `None` as a map value generates a random port;
+                      Dicts are ordered. By convention the first key-val pair is designated to the HTTP interface.
+                      Example: {4200: None, 5432: 15431} - port 4200 inside the container will be mapped
```
```suggestion
                      Example: {4200: None, 5432: 15432} - port 4200 inside the container will be mapped
```

pilosus replied:
Fixed

matriv commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,87 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
+        :param user:  optional username to access the DB; if None, try respective environment variable
+        :param password: optional password to access the DB; if None, try respective environment variable
+        :param dbname: optional database name to access the DB; if None, try respective environment variable
+        :param dialect: a string with the dialect to generate a DB URI
```
why do we want to expose `dialect` as parameter? is there a use case to change it?

amotl replied:
I am not 100% sure, but I think when passed `"http"` instead of `None`, the corresponding utility function will return the HTTP-based connection URL, and not the SQLAlchemy-based connection URL, in the same manner as `get_http_url()` does it.

The `dialect` argument has already been part of the implementation before the refactoring, and I think it is part of the official/designated constructor signature. See [^1].

[^1]: https://github.com/crate-workbench/cratedb-toolkit/blob/v0.0.2/cratedb_toolkit/testing/testcontainers/cratedb.py#L61

matriv replied:
`dialect` is normally used to use a DB specific dialect instead of a more generic one, e.g. psql vs jdbc, it's what we use to set `crate` instead of `psql` for the cases that we want to use the `crate` jdbc driver instead of the psql one. `http` vs binary `psql` protocol is not a dialect, but a protocol.

matriv replied:
Not saying that it's part of the PR, just mentioning that it seems weird to me to expose a "dialect".

pilosus replied:
Actually, I've double checked the dialect thing and you are right @matriv - no need to have it in the class constructor. It's enough to have it in the `get_connection_url` as an argument with the default value `crate`. Gonna fix that

matriv commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,87 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
+                      Example: {4200: None, 5432: 15431} - port 4200 inside the container will be mapped
+                      to a random port on the host, internal port 5432 for PSQL interface will be mapped
+                      to the 15432 port on the host.
+        :param user:  optional username to access the DB; if None, try respective environment variable
```
here and below, maybe replace `respective` with the actual name of the env variable.

pilosus replied:
Fixed

matriv reviewed (commented):
thanks for your work @pilosus! I've left a few comments.

matriv reviewed (approved):
thank you @pilosus !

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,91 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
     KEEPALIVE = asbool(os.environ.get("CRATEDB_KEEPALIVE", os.environ.get("TC_KEEPALIVE", False)))
+    CMD_OPTS = {
+        "discovery.type": "single-node",
+        "node.attr.storage": "hot",
```
Yeah, well spotted. It is indeed a configuration detail which is specific to the test cases for `cratedb_toolkit.retention`. 

@pilosus: Do you think we can improve this spot, so that corresponding configuration settings can be defined on behalf of the snippet in `conftest.py`? This time, it will probably not be so easy, because the test adapter will already need this information at startup time.

amotl replied:
> This time, it will probably not be so easy, because the test adapter will already need this information at startup time. Maybe you have an idea how to handle this elegantly?
>
> While being at it: Of course, it would not just be about the specific `node.attr.storage` parameter, but about any other parameters as well.

Ah, I see there is already `cmd_opts` on the constructor. So, `{"node.attr.storage": "hot"}` would just need to be moved over there.

pilosus replied:
Exactly. And given that we use [dict merge](https://github.com/crate-workbench/cratedb-toolkit/pull/82/files#diff-a269e42ec7e05ba66ed28ac184deeaa00fb81a0db3ccb47d34635c654f1b68bcR95), you can even override the default settings

```
In [1]: from cratedb_toolkit.testing.testcontainers.cratedb  import CrateDBContainer

In [2]: c = CrateDBContainer(cmd_opts={"node.attr.storage": "cold"})

In [3]: c._command
Out[3]: '-Cdiscovery.type=single-node -Cnode.attr.storage=cold -Cpath.repo=/tmp/snapshots'
```

I'll remove the `node.attr.storage` from the default though

amotl replied:
Thanks. What do you think about `path.repo`, @seut? It could be convenient for testing to have it configured by default.

pilosus replied:
@seut @amotl hmm, we I delete `node.attr.storage=hot`
tests: 
- `tests/retention/test_cli.py::test_run_delete_basic` 
- `tests/retention/test_cli.py::test_run_delete_dryrun`
- `tests/retention/test_cli.py::test_run_reallocate`
hang.

Here are the logs:
```
2023-11-24 17:39:14,494 [cratedb_toolkit.retention.setup.schema] INFO    : Installing retention policy bookkeeping table at database 'crate://crate:REDACTED@localhost:33018', table TableAddress(schema='testdrive-ext', table='retention_policy')
2023-11-24 17:39:14,902 [cratedb_toolkit.retention.store     ] INFO    : Connecting to database crate://crate:REDACTED@localhost:33018, table "testdrive-ext"."retention_policy"
Waiting to be ready...
2023-11-24 17:39:15,366 [testcontainers.core.waiting_utils   ] INFO    : Waiting to be ready...
```

Interestingly enough, `test_run_delete_with_tags_match` works well

amotl replied:
Yeah, the test cases defined for `cratedb_toolkit.retention` need this setting to be configured, so it will need to go into `tests/conftest.py` somehow. On the other hand, it should not be part of the generic configuration. That's yet another cliff we need to take.

pilosus replied:
Given the whole chain of the nested fixtures with different scopes, it will either require some more time from me next week, or a simpler solution with `cratedb` override for `tests/retention/conftest.py` that have `node.attr.storage=hot` in it.

amotl replied:
We can easily handle that on a later iteration, and/or discuss possible solutions beforehand.

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,91 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
-        self._command += " -Cpath.repo=/tmp/snapshots"
+        self._name = "testcontainers-cratedb"
+
+        cmd_opts = cmd_opts if cmd_opts else {}
```
Just a nit: I usually write this like that. Is it a good idea?

```suggestion
        cmd_opts = cmd_opts or {}
```

pilosus replied:
Done

amotl commented on cratedb_toolkit/testing/testcontainers/cratedb.py:
```diff
@@ -51,44 +57,91 @@ class CrateDBContainer(KeepaliveContainer, DbContainer):
+        for key, val in opts.items():
+            if isinstance(val, bool):
+                val = str(val).lower()
+            cmd.append("-C{}={}".format(key, val))
```
Is there a reason not to use f-strings already?
```suggestion
            cmd.append(f"-C{key}={val}")
```

pilosus replied:
Fixed.
Alhough I personally not a big fan of f-strings, because sometimes people abuse their flexibility with larger expressions and, thus, damage readability. But it's a personal choice rather than anything else.

amotl reviewed (commented):
Hi Vitaly. Thanks once again for your work on this. I've added two more nitpicky suggestions ;].
